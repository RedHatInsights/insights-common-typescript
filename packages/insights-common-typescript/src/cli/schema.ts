#!/usr/bin/env node

import { Command } from 'commander';
import * as jq from 'node-jq';
import { CLIEngine } from 'eslint';
import fetch from 'node-fetch';
import prettier from 'prettier';
import isUrl from 'is-url';
import { existsSync, lstatSync, writeFileSync } from 'fs';

import { SchemaActionBuilder, SchemaTypeBuilder, OpenAPI3 } from './schema/';
import { buildApiDescriptor } from './schema/ApiDescriptorBuilder';

export interface Options {
    inputFile: string;
    output: string;
    skipPostProcess?: boolean;
    skipActionGenerator?: boolean;
}

const getProgram = () => {
    const program = new Command();

    program
    .description('Generates the actions needed by react-fetching-library to do the queries out of the openapi.json file')
    .option(
        '-s, --skip-post-process',
        'Skips the postprocess (prettier and linter)',
        false
    )
    .option(
        '-sa, --skip-action-generator',
        'Skips the actions generators used for react-fetching-library',
        false
    )
    .requiredOption(
        '-i, --input-file <openapijson-path>',
        'URL or local path to the openapi.json file.'
    )
    .requiredOption(
        '-o, --output <output-path>',
        'Output file to put the generated files.'
    );

    return program;
};

export const execute = async (options: Options) => {

    let inputType = 'file';
    let input = options.inputFile;

    if (isUrl(options.inputFile)) {
        inputType = 'string';
        input = await fetch(options.inputFile, {
            headers: {
                Accept: 'application/json'
            }
        }).then(res => res.text());
    }

    if (existsSync(options.output)) {
        const fileStat = lstatSync(options.output);
        if (fileStat.isDirectory()) {
            options.output = `${options.output}/Generated.ts`;
        }
    }

    return jq.run('.', input, {
        sort: true,
        input: inputType
    }).then(output => JSON.parse(output as string) as OpenAPI3)
    .then(openapi => buildApiDescriptor(openapi, {
        nonRequiredPropertyIsNull: true
    }))
    .then(descriptor => {
        const actionGeneratorHeaders = [
            'import { actionBuilder, ValidatedResponse, ActionValidatable } from \'@redhat-cloud-services/insights-common-typescript\';\n',
            'import { Action } from \'react-fetching-library\';\n'
        ];

        const buffer: Array<string> = [
            '/**\n',
            '* Generated code, DO NOT modify directly.\n',
            '*/\n',
            'import * as z from \'zod\';\n',
            ...(options.skipActionGenerator ? [] : actionGeneratorHeaders),
            '/* eslint-disable @typescript-eslint/camelcase */\n',
            '/* eslint-disable @typescript-eslint/no-use-before-define */\n\n'
        ];

        const typeBuilder = new SchemaTypeBuilder(descriptor, buffer);
        typeBuilder.build();

        if (!options.skipActionGenerator) {
            const actionBuilder = new SchemaActionBuilder(descriptor, buffer);
            actionBuilder.build();
        }

        return buffer.join('');
    }).then(content => {
        if (options.skipPostProcess) {
            return content;
        }

        return prettier.format(
            content,
            {
                parser: 'typescript'
            }
        );
    }).then(async (content) => {
        if (options.skipPostProcess) {
            return writeFileSync(options.output, content);
        }

        const eslint = new CLIEngine({
            fix: true
        });
        const results = await eslint.executeOnText(content, options.output);
        if (results.fixableErrorCount || results.fixableWarningCount) {
            return CLIEngine.outputFixes(results);
        }

        return writeFileSync(options.output, content);
    });
};

if (require.main === module) {
    const program = getProgram();
    program.parse(process.argv);
    execute(program as unknown as Options);
}
