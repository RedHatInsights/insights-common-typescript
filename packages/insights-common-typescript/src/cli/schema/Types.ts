import { StringMap } from './types/Helpers';
import { SchemaWithTypeName } from './types/Schemas';

export type Buffer = Array<string>;

export interface APIDescriptor {
    path: string;
    components: Components;
}

export interface Components {
    schemas?: StringMap<SchemaWithTypeName>;
}
