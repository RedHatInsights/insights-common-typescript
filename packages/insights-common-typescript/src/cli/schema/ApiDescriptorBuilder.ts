import { isReference, OpenAPI3 } from './types/OpenAPI3';
import { APIDescriptor } from './Types';
import { StringMap } from './types/Helpers';
import { Schema, SchemaObject, SchemaType, SchemaUnknown, SchemaWithTypeName, Type } from './types/Schemas';

const refToName = (reference: OpenAPI3.Reference) => {
    const [ last ] = reference.$ref.split('/').reverse();
    return last;
};

export interface ApiDescriptorBuilderOptions {
    nonRequiredPropertyIsNull: boolean;
}

class ApiDescriptorBuilder {
    readonly types: StringMap<Type>;
    readonly openapi: Readonly<OpenAPI3>;
    readonly options: ApiDescriptorBuilderOptions;

    constructor(openapi: OpenAPI3, options: Partial<ApiDescriptorBuilderOptions>) {
        this.openapi = openapi;
        this.types = this.getTypesPlaceholders();
        this.options = {
            nonRequiredPropertyIsNull: false,
            ...options
        };
    }

    public build(): APIDescriptor {
        return {
            path: this.getPath(),
            components: this.getSchemaComponents()
        };
    }

    private getSchemaComponents(): StringMap<SchemaWithTypeName> {
        const schemas: StringMap<SchemaWithTypeName> = {};
        if (this.openapi.components?.schemas) {
            for (const [ typeName, schema ] of Object.entries(this.openapi.components.schemas)) {
                if (isReference(schema)) {
                    throw new Error('Invalid reference found at component level');
                } else {
                    schemas[typeName] = {
                        typeName,
                        ...this.getSchema(schema) as Schema
                    };
                }
            }
        }

        return schemas;
    }

    private getSchemaType(schema: OpenAPI3.Schema): Schema {
        if (schema.allOf) {
            return {
                type: SchemaType.ALL_OF,
                allOf: schema.allOf.map(s => this.getSchema(s))
            };
        } else if (schema.oneOf) {
            return {
                type: SchemaType.ONE_OF,
                oneOf: schema.oneOf.map(s => this.getSchema(s))
            };
        } else if (schema.anyOf) {
            return {
                type: SchemaType.ANY_OF,
                anyOf: schema.anyOf.map(s => this.getSchema(s))
            };
        } else if (schema.enum) {
            return {
                type: SchemaType.ENUM,
                enum: schema.enum
            };
        } else if (schema.type) {
            switch (schema.type) {
                case 'array':
                    return {
                        type: SchemaType.ARRAY,
                        items: schema.items ? this.getSchema(schema.items) : {
                            type: SchemaType.UNKNOWN
                        }
                    };
                case 'number':
                    return {
                        type: SchemaType.NUMBER
                    };
                case 'integer':
                    return {
                        type: SchemaType.INTEGER
                    };
                case 'string':
                    return {
                        type: SchemaType.STRING
                    };
                case 'boolean':
                    return {
                        type: SchemaType.BOOLEAN
                    };
                case 'null':
                    return {
                        type: SchemaType.NULL
                    };
                case 'object':
                    return this.getSchemaForObject({ ...schema, type: 'object' });
                default:
                    throw new Error(`Unknown type found: ${schema.type} for schema ${JSON.stringify(schema)}`);
            }
        } else {
            return {
                type: SchemaType.UNKNOWN
            };
        }
    }

    private getSchema(schemaOrRef: OpenAPI3.Schema | OpenAPI3.Reference): Schema | Type {
        if (isReference(schemaOrRef)) {
            const typeName = refToName(schemaOrRef);
            return this.types[typeName];
        } else {
            const schema = this.getSchemaType(schemaOrRef);
            if (schemaOrRef.nullable) {
                schema.isNullable = true;
            }

            return schema;
        }
    }

    private getSchemaForObject(schema: OpenAPI3.Schema & { type: 'object' }): SchemaObject | SchemaUnknown {
        if (schema.properties || schema.additionalProperties) {
            let additionalProperties: SchemaObject['additionalProperties'] = undefined;
            if (schema.additionalProperties) {
                if (schema.additionalProperties === true) {
                    additionalProperties = {
                        type: SchemaType.UNKNOWN
                    };
                } else {
                    additionalProperties = this.getSchema(schema.additionalProperties);
                }
            }

            let properties: SchemaObject['properties'] = undefined;
            if (schema.properties) {
                properties = {};
                for (const [ key, value ] of Object.entries(schema.properties)) {
                    properties[key] = this.getSchema(value);
                    if (schema.required?.includes(key)) {
                        properties[key].isOptional = true;
                        if (this.options.nonRequiredPropertyIsNull) {
                            properties[key].isNullable = true;
                        }
                    }
                }
            }

            return {
                type: SchemaType.OBJECT,
                additionalProperties,
                properties
            };
        }

        return {
            type: SchemaType.UNKNOWN
        };
    }

    private getPath(): string {
        if (this.openapi.servers) {
            const variables = this.openapi.servers[0].variables;
            if (variables?.basePath?.default) {
                return variables.basePath.default;
            }
        }

        return '';
    }

    private getTypesPlaceholders(): StringMap<Type> {
        const types: StringMap<Type> = {};
        if (this.openapi.components?.schemas) {
            for (const [ typeName ] of Object.entries(this.openapi.components.schemas)) {
                types[typeName] = {
                    typeName,
                    schema: null
                } as unknown as Type;
            }
        }

        return types;
    }
}

export const buildApiDescriptor = (openapi: Readonly<OpenAPI3>, options: Partial<ApiDescriptorBuilderOptions>): APIDescriptor => {
    const builder = new ApiDescriptorBuilder(openapi, options);
    return builder.build();
};
