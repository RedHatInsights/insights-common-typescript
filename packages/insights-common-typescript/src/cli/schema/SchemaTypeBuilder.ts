import { SchemaBase } from './SchemaBase';
import { Buffer, OpenAPI3, StringMap } from './Types';

export class SchemaTypeBuilder extends SchemaBase {

    constructor(openapi: OpenAPI3, buffer: Buffer) {
        super(openapi, buffer);
    }

    public build() {
        if (this.openapi?.components?.schemas) {
            const schemas = this.openapi.components.schemas;
            this.types(schemas);
            this.functionTypes(schemas);
        }
    }

    private types(schemas: StringMap<OpenAPI3.Schema | OpenAPI3.Reference>) {
        for (const [ typeName ] of Object.entries(schemas)) {
            this.append(`export const ${this.typeName(typeName)} = ${this.functionName(typeName)}();\n`);
            this.append(`export type ${this.typeName(typeName)} = z.infer<typeof ${this.typeName(typeName)}>;\n`);
            this.append('\n');
        }

        this.append('\n');
    }

    private functionTypes(schemas: StringMap<OpenAPI3.Schema | OpenAPI3.Reference>) {
        for (const [ typeName, schema ] of Object.entries(schemas)) {
            this.append(`export function ${this.functionName(typeName)}() {\nreturn `);
            // Making top-level schemas required unless `required` is defined
            this.schema(schema);
            this.append(';\n}\n\n');
        }
    }
}
