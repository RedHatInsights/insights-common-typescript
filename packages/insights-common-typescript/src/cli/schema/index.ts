export { SchemaActionBuilder } from './SchemaActionBuilder';
export { SchemaTypeBuilder } from './SchemaTypeBuilder';
export { OpenAPI3 } from './types/OpenAPI3';
