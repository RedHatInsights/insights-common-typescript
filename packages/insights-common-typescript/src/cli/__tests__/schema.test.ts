import fetchMockJest from 'fetch-mock-jest';
import { execute } from '../schema';
import { existsSync, readFileSync, mkdirSync } from 'fs';
import rimRaf from 'rimraf';
import fetchMock from 'node-fetch';

jest.mock('node-fetch', () => fetchMockJest.sandbox());

describe.each([
    './src/cli/__tests__/notifications-openapi.json',
    './src/cli/__tests__/policies-openapi.json',
    './src/cli/__tests__/simple-openapi.json'
])('src/cli/schema for %s', (filename) => {

    const tempSchemaDir = './tmp/schemas';

    beforeEach(() => {
        mkdirSync(tempSchemaDir, {
            recursive: true
        });
    });

    afterEach(() => {
        rimRaf.sync(tempSchemaDir);
    });

    it('execute input file accepts path', () => {
        return execute({
            inputFile: filename,
            output: tempSchemaDir,
            skipPostProcess: false,
            addEslintIgnore: true,
            skipActionGenerator: false
        }).then(() => {
            expect(existsSync(`${tempSchemaDir}/Generated.ts`)).toBeTruthy();
            expect(readFileSync(`${tempSchemaDir}/Generated.ts`).toString()).toMatchSnapshot();

        });
    });

    it('execute accepts urls', async () => {
        (fetchMock as any).get('http://foobar.baz/my-openapi.json', {
            body: readFileSync(filename).toString(),
            status: 200
        });

        return execute({
            inputFile: 'http://foobar.baz/my-openapi.json',
            output: tempSchemaDir,
            addEslintIgnore: true,
            skipPostProcess: false,
            skipActionGenerator: false
        }).then(() => {
            (fetchMock as any).restore();
            expect(existsSync(`${tempSchemaDir}/Generated.ts`)).toBeTruthy();
            expect(readFileSync(`${tempSchemaDir}/Generated.ts`).toString()).toMatchSnapshot();
        });
    });
});
