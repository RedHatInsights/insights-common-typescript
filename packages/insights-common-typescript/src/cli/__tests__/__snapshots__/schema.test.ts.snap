// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/cli/schema execute accepts urls 1`] = `
"/**
* Generated code, DO NOT modify directly.
*/
/* eslint-disable @typescript-eslint/no-empty-interface */

export interface Fact {
  id?: number;
  name?: string;
  type?: FactType;
}

export enum FactType {
  BOOLEAN = 'BOOLEAN',
  INT = 'INT',
  LIST = 'LIST',
  STRING = 'STRING',
}

export interface HistoryItem {
  ctime?: number;
  hostName?: string;
  id?: string;
}

export type List = Array<{
}>;

export type ListHistoryItem = Array<HistoryItem>;

export type ListPolicy = Array<Policy>;

export type ListUUID = Array<string>;

export interface MapStringString {
  [key: string]: string;
}

export interface Meta {
  count?: number;
}

export interface PagedResponseOfHistoryItem {
  data?: ListHistoryItem;
  links?: MapStringString;
  meta?: Meta;
}

export interface PagedResponseOfPolicy {
  data?: ListPolicy;
  links?: MapStringString;
  meta?: Meta;
}

export interface Policy {
  actions?: string;
  conditions: string;
  ctime?: string;
  description?: string;
  id?: Uuid;
  isEnabled?: boolean;
  lastTriggered?: number;
  mtime?: string;
  name: string;
}

export type Uuid = string;

"
`;

exports[`src/cli/schema execute accepts urls 2`] = `
"/**
* Generated code, DO NOT modify directly.
*/
/* eslint-disable max-len */

import { actionBuilder } from '@redhat-cloud-services/insights-common-typescript';
import { Action } from 'react-fetching-library';
import * as schemas from './Types';

export interface UseGetPoliciesParams {
    filterOpDescription?: 'equal' | 'like' | 'ilike' | 'not_equal';
    filterOpName?: 'equal' | 'like' | 'ilike' | 'not_equal';
    filterDescription?: string;
    filterIsEnabled?: 'true' | 'false';
    filterName?: string;
    limit?: number;
    offset?: number;
    sortColumn?: 'name' | 'description' | 'is_enabled' | 'mtime';
    sortDirection?: 'asc' | 'desc';
}

export interface UsePostPoliciesParams {
    body?: schemas.Policy;
    alsoStore?: boolean;
}

export interface UseGetPoliciesIdsParams {
    filterOpDescription?: 'equal' | 'like' | 'ilike' | 'not_equal';
    filterOpName?: 'equal' | 'like' | 'ilike' | 'not_equal';
    filterDescription?: string;
    filterIsEnabled?: 'true' | 'false';
    filterName?: string;
}

export interface UseDeletePoliciesIdsParams {
    body?: schemas.ListUUID;
}

export interface UsePostPoliciesIdsEnabledParams {
    body?: schemas.ListUUID;
    enabled?: boolean;
}

export interface UsePostPoliciesValidateParams {
    body?: schemas.Policy;
}

export interface UsePostPoliciesValidateNameParams {
    body?: string;
    id?: schemas.Uuid;
}

export interface UseGetPoliciesByIdParams {
    id: schemas.Uuid;
}

export interface UseDeletePoliciesByIdParams {
    id: schemas.Uuid;
}

export interface UsePostPoliciesByIdEnabledParams {
    id: schemas.Uuid;
    enabled?: boolean;
}

export interface UseGetPoliciesByIdHistoryTriggerParams {
    id: schemas.Uuid;
    filterOpId?: 'equal' | 'not_equal' | 'like';
    filterOpName?: 'equal' | 'like' | 'not_equal';
    filterId?: string;
    filterName?: string;
    limit?: number;
    offset?: number;
    sortColumn?: 'hostName' | 'id' | 'ctime';
    sortDirection?: 'asc' | 'desc';
}

export interface UsePutPoliciesByPolicyIdParams {
    body?: schemas.Policy;
    policyId: schemas.Uuid;
    dry?: boolean;
}

/** Just a filler to have a defined return code for the base path */
export const actionGet = (): Action<unknown> => {
    const path = '/api/policies/v1.0/';

    const query = {} as Record<string, any>;

    return actionBuilder('GET', path)
    .queryParams(query)
    .build();
};

/** Retrieve a list of fact (keys) along with their data types */
export const actionGetFacts = (): Action<Array<schemas.Fact> | unknown> => {
    const path = '/api/policies/v1.0/facts';

    const query = {} as Record<string, any>;

    return actionBuilder('GET', path)
    .queryParams(query)
    .build();
};

/** Return all policies for a given account */
export const actionGetPolicies = (params: UseGetPoliciesParams): Action<schemas.PagedResponseOfPolicy | unknown> => {
    const path = '/api/policies/v1.0/policies';

    const query = {} as Record<string, any>;
    query['filter:op[description]'] = params.filterOpDescription;
    query['filter:op[name]'] = params.filterOpName;
    query['filter[description]'] = params.filterDescription;
    query['filter[is_enabled]'] = params.filterIsEnabled;
    query['filter[name]'] = params.filterName;
    query.limit = params.limit;
    query.offset = params.offset;
    query.sortColumn = params.sortColumn;
    query.sortDirection = params.sortDirection;

    return actionBuilder('GET', path)
    .queryParams(query)
    .build();
};

/** Validate (and possibly persist) a passed policy for the given account */
export const actionPostPolicies = (params: UsePostPoliciesParams): Action<schemas.Policy | {
  msg?: string;
} | {
  msg?: string;
} | unknown> => {
    const path = '/api/policies/v1.0/policies';

    const query = {} as Record<string, any>;
    query.alsoStore = params.alsoStore;

    return actionBuilder('POST', path)
    .queryParams(query)
    .data(params.body)
    .build();
};

/** Return all policy ids for a given account after applying the filters */
export const actionGetPoliciesIds = (params: UseGetPoliciesIdsParams): Action<schemas.List | unknown> => {
    const path = '/api/policies/v1.0/policies/ids';

    const query = {} as Record<string, any>;
    query['filter:op[description]'] = params.filterOpDescription;
    query['filter:op[name]'] = params.filterOpName;
    query['filter[description]'] = params.filterDescription;
    query['filter[is_enabled]'] = params.filterIsEnabled;
    query['filter[name]'] = params.filterName;

    return actionBuilder('GET', path)
    .queryParams(query)
    .build();
};

/** Delete policies for a customer by the ids passed in the body. Result will be a list of deleted UUIDs */
export const actionDeletePoliciesIds = (params: UseDeletePoliciesIdsParams): Action<Array<schemas.Uuid> | unknown> => {
    const path = '/api/policies/v1.0/policies/ids';

    const query = {} as Record<string, any>;

    return actionBuilder('DELETE', path)
    .queryParams(query)
    .data(params.body)
    .build();
};

/** Enable/disable policies identified by list of uuid in body */
export const actionPostPoliciesIdsEnabled = (params: UsePostPoliciesIdsEnabledParams): Action<unknown> => {
    const path = '/api/policies/v1.0/policies/ids/enabled';

    const query = {} as Record<string, any>;
    query.enabled = params.enabled;

    return actionBuilder('POST', path)
    .queryParams(query)
    .data(params.body)
    .build();
};

/** Validates a Policy condition */
export const actionPostPoliciesValidate = (params: UsePostPoliciesValidateParams): Action<unknown> => {
    const path = '/api/policies/v1.0/policies/validate';

    const query = {} as Record<string, any>;

    return actionBuilder('POST', path)
    .queryParams(query)
    .data(params.body)
    .build();
};

/** Validates the Policy.name and verifies if it is unique. */
export const actionPostPoliciesValidateName = (params: UsePostPoliciesValidateNameParams): Action<unknown> => {
    const path = '/api/policies/v1.0/policies/validate-name';

    const query = {} as Record<string, any>;
    query.id = params.id;

    return actionBuilder('POST', path)
    .queryParams(query)
    .data(params.body)
    .build();
};

/** Retrieve a single policy for a customer by its id */
export const actionGetPoliciesById = (params: UseGetPoliciesByIdParams): Action<schemas.Policy | unknown> => {
    const path = '/api/policies/v1.0/policies/{id}'
    .replace('{id}', params.id.toString());

    const query = {} as Record<string, any>;

    return actionBuilder('GET', path)
    .queryParams(query)
    .build();
};

/** Delete a single policy for a customer by its id */
export const actionDeletePoliciesById = (params: UseDeletePoliciesByIdParams): Action<unknown> => {
    const path = '/api/policies/v1.0/policies/{id}'
    .replace('{id}', params.id.toString());

    const query = {} as Record<string, any>;

    return actionBuilder('DELETE', path)
    .queryParams(query)
    .build();
};

/** Enable/disable a policy */
export const actionPostPoliciesByIdEnabled = (params: UsePostPoliciesByIdEnabledParams): Action<unknown> => {
    const path = '/api/policies/v1.0/policies/{id}/enabled'
    .replace('{id}', params.id.toString());

    const query = {} as Record<string, any>;

    query.enabled = params.enabled;

    return actionBuilder('POST', path)
    .queryParams(query)
    .build();
};

/** Retrieve the trigger history of a single policy */
export const actionGetPoliciesByIdHistoryTrigger = (params: UseGetPoliciesByIdHistoryTriggerParams): Action<schemas.PagedResponseOfHistoryItem | unknown> => {
    const path = '/api/policies/v1.0/policies/{id}/history/trigger'
    .replace('{id}', params.id.toString());

    const query = {} as Record<string, any>;

    query['filter:op[id]'] = params.filterOpId;
    query['filter:op[name]'] = params.filterOpName;
    query['filter[id]'] = params.filterId;
    query['filter[name]'] = params.filterName;
    query.limit = params.limit;
    query.offset = params.offset;
    query.sortColumn = params.sortColumn;
    query.sortDirection = params.sortDirection;

    return actionBuilder('GET', path)
    .queryParams(query)
    .build();
};

/** Update a single policy for a customer by its id */
export const actionPutPoliciesByPolicyId = (params: UsePutPoliciesByPolicyIdParams): Action<{
  msg?: string;
} | unknown> => {
    const path = '/api/policies/v1.0/policies/{policyId}'
    .replace('{policyId}', params.policyId.toString());

    const query = {} as Record<string, any>;

    query.dry = params.dry;

    return actionBuilder('PUT', path)
    .queryParams(query)
    .data(params.body)
    .build();
};

export const actionGetStatus = (): Action<unknown> => {
    const path = '/api/policies/v1.0/status';

    const query = {} as Record<string, any>;

    return actionBuilder('GET', path)
    .queryParams(query)
    .build();
};

"
`;

exports[`src/cli/schema execute input file accepts path 1`] = `
"/**
* Generated code, DO NOT modify directly.
*/
/* eslint-disable @typescript-eslint/no-empty-interface */

export interface Fact {
  id?: number;
  name?: string;
  type?: FactType;
}

export enum FactType {
  BOOLEAN = 'BOOLEAN',
  INT = 'INT',
  LIST = 'LIST',
  STRING = 'STRING',
}

export interface HistoryItem {
  ctime?: number;
  hostName?: string;
  id?: string;
}

export type List = Array<{
}>;

export type ListHistoryItem = Array<HistoryItem>;

export type ListPolicy = Array<Policy>;

export type ListUUID = Array<string>;

export interface MapStringString {
  [key: string]: string;
}

export interface Meta {
  count?: number;
}

export interface PagedResponseOfHistoryItem {
  data?: ListHistoryItem;
  links?: MapStringString;
  meta?: Meta;
}

export interface PagedResponseOfPolicy {
  data?: ListPolicy;
  links?: MapStringString;
  meta?: Meta;
}

export interface Policy {
  actions?: string;
  conditions: string;
  ctime?: string;
  description?: string;
  id?: Uuid;
  isEnabled?: boolean;
  lastTriggered?: number;
  mtime?: string;
  name: string;
}

export type Uuid = string;

"
`;

exports[`src/cli/schema execute input file accepts path 2`] = `
"/**
* Generated code, DO NOT modify directly.
*/
/* eslint-disable max-len */

import { actionBuilder } from '@redhat-cloud-services/insights-common-typescript';
import { Action } from 'react-fetching-library';
import * as schemas from './Types';

export interface UseGetPoliciesParams {
    filterOpDescription?: 'equal' | 'like' | 'ilike' | 'not_equal';
    filterOpName?: 'equal' | 'like' | 'ilike' | 'not_equal';
    filterDescription?: string;
    filterIsEnabled?: 'true' | 'false';
    filterName?: string;
    limit?: number;
    offset?: number;
    sortColumn?: 'name' | 'description' | 'is_enabled' | 'mtime';
    sortDirection?: 'asc' | 'desc';
}

export interface UsePostPoliciesParams {
    body?: schemas.Policy;
    alsoStore?: boolean;
}

export interface UseGetPoliciesIdsParams {
    filterOpDescription?: 'equal' | 'like' | 'ilike' | 'not_equal';
    filterOpName?: 'equal' | 'like' | 'ilike' | 'not_equal';
    filterDescription?: string;
    filterIsEnabled?: 'true' | 'false';
    filterName?: string;
}

export interface UseDeletePoliciesIdsParams {
    body?: schemas.ListUUID;
}

export interface UsePostPoliciesIdsEnabledParams {
    body?: schemas.ListUUID;
    enabled?: boolean;
}

export interface UsePostPoliciesValidateParams {
    body?: schemas.Policy;
}

export interface UsePostPoliciesValidateNameParams {
    body?: string;
    id?: schemas.Uuid;
}

export interface UseGetPoliciesByIdParams {
    id: schemas.Uuid;
}

export interface UseDeletePoliciesByIdParams {
    id: schemas.Uuid;
}

export interface UsePostPoliciesByIdEnabledParams {
    id: schemas.Uuid;
    enabled?: boolean;
}

export interface UseGetPoliciesByIdHistoryTriggerParams {
    id: schemas.Uuid;
    filterOpId?: 'equal' | 'not_equal' | 'like';
    filterOpName?: 'equal' | 'like' | 'not_equal';
    filterId?: string;
    filterName?: string;
    limit?: number;
    offset?: number;
    sortColumn?: 'hostName' | 'id' | 'ctime';
    sortDirection?: 'asc' | 'desc';
}

export interface UsePutPoliciesByPolicyIdParams {
    body?: schemas.Policy;
    policyId: schemas.Uuid;
    dry?: boolean;
}

/** Just a filler to have a defined return code for the base path */
export const actionGet = (): Action<unknown> => {
    const path = '/api/policies/v1.0/';

    const query = {} as Record<string, any>;

    return actionBuilder('GET', path)
    .queryParams(query)
    .build();
};

/** Retrieve a list of fact (keys) along with their data types */
export const actionGetFacts = (): Action<Array<schemas.Fact> | unknown> => {
    const path = '/api/policies/v1.0/facts';

    const query = {} as Record<string, any>;

    return actionBuilder('GET', path)
    .queryParams(query)
    .build();
};

/** Return all policies for a given account */
export const actionGetPolicies = (params: UseGetPoliciesParams): Action<schemas.PagedResponseOfPolicy | unknown> => {
    const path = '/api/policies/v1.0/policies';

    const query = {} as Record<string, any>;
    query['filter:op[description]'] = params.filterOpDescription;
    query['filter:op[name]'] = params.filterOpName;
    query['filter[description]'] = params.filterDescription;
    query['filter[is_enabled]'] = params.filterIsEnabled;
    query['filter[name]'] = params.filterName;
    query.limit = params.limit;
    query.offset = params.offset;
    query.sortColumn = params.sortColumn;
    query.sortDirection = params.sortDirection;

    return actionBuilder('GET', path)
    .queryParams(query)
    .build();
};

/** Validate (and possibly persist) a passed policy for the given account */
export const actionPostPolicies = (params: UsePostPoliciesParams): Action<schemas.Policy | {
  msg?: string;
} | {
  msg?: string;
} | unknown> => {
    const path = '/api/policies/v1.0/policies';

    const query = {} as Record<string, any>;
    query.alsoStore = params.alsoStore;

    return actionBuilder('POST', path)
    .queryParams(query)
    .data(params.body)
    .build();
};

/** Return all policy ids for a given account after applying the filters */
export const actionGetPoliciesIds = (params: UseGetPoliciesIdsParams): Action<schemas.List | unknown> => {
    const path = '/api/policies/v1.0/policies/ids';

    const query = {} as Record<string, any>;
    query['filter:op[description]'] = params.filterOpDescription;
    query['filter:op[name]'] = params.filterOpName;
    query['filter[description]'] = params.filterDescription;
    query['filter[is_enabled]'] = params.filterIsEnabled;
    query['filter[name]'] = params.filterName;

    return actionBuilder('GET', path)
    .queryParams(query)
    .build();
};

/** Delete policies for a customer by the ids passed in the body. Result will be a list of deleted UUIDs */
export const actionDeletePoliciesIds = (params: UseDeletePoliciesIdsParams): Action<Array<schemas.Uuid> | unknown> => {
    const path = '/api/policies/v1.0/policies/ids';

    const query = {} as Record<string, any>;

    return actionBuilder('DELETE', path)
    .queryParams(query)
    .data(params.body)
    .build();
};

/** Enable/disable policies identified by list of uuid in body */
export const actionPostPoliciesIdsEnabled = (params: UsePostPoliciesIdsEnabledParams): Action<unknown> => {
    const path = '/api/policies/v1.0/policies/ids/enabled';

    const query = {} as Record<string, any>;
    query.enabled = params.enabled;

    return actionBuilder('POST', path)
    .queryParams(query)
    .data(params.body)
    .build();
};

/** Validates a Policy condition */
export const actionPostPoliciesValidate = (params: UsePostPoliciesValidateParams): Action<unknown> => {
    const path = '/api/policies/v1.0/policies/validate';

    const query = {} as Record<string, any>;

    return actionBuilder('POST', path)
    .queryParams(query)
    .data(params.body)
    .build();
};

/** Validates the Policy.name and verifies if it is unique. */
export const actionPostPoliciesValidateName = (params: UsePostPoliciesValidateNameParams): Action<unknown> => {
    const path = '/api/policies/v1.0/policies/validate-name';

    const query = {} as Record<string, any>;
    query.id = params.id;

    return actionBuilder('POST', path)
    .queryParams(query)
    .data(params.body)
    .build();
};

/** Retrieve a single policy for a customer by its id */
export const actionGetPoliciesById = (params: UseGetPoliciesByIdParams): Action<schemas.Policy | unknown> => {
    const path = '/api/policies/v1.0/policies/{id}'
    .replace('{id}', params.id.toString());

    const query = {} as Record<string, any>;

    return actionBuilder('GET', path)
    .queryParams(query)
    .build();
};

/** Delete a single policy for a customer by its id */
export const actionDeletePoliciesById = (params: UseDeletePoliciesByIdParams): Action<unknown> => {
    const path = '/api/policies/v1.0/policies/{id}'
    .replace('{id}', params.id.toString());

    const query = {} as Record<string, any>;

    return actionBuilder('DELETE', path)
    .queryParams(query)
    .build();
};

/** Enable/disable a policy */
export const actionPostPoliciesByIdEnabled = (params: UsePostPoliciesByIdEnabledParams): Action<unknown> => {
    const path = '/api/policies/v1.0/policies/{id}/enabled'
    .replace('{id}', params.id.toString());

    const query = {} as Record<string, any>;

    query.enabled = params.enabled;

    return actionBuilder('POST', path)
    .queryParams(query)
    .build();
};

/** Retrieve the trigger history of a single policy */
export const actionGetPoliciesByIdHistoryTrigger = (params: UseGetPoliciesByIdHistoryTriggerParams): Action<schemas.PagedResponseOfHistoryItem | unknown> => {
    const path = '/api/policies/v1.0/policies/{id}/history/trigger'
    .replace('{id}', params.id.toString());

    const query = {} as Record<string, any>;

    query['filter:op[id]'] = params.filterOpId;
    query['filter:op[name]'] = params.filterOpName;
    query['filter[id]'] = params.filterId;
    query['filter[name]'] = params.filterName;
    query.limit = params.limit;
    query.offset = params.offset;
    query.sortColumn = params.sortColumn;
    query.sortDirection = params.sortDirection;

    return actionBuilder('GET', path)
    .queryParams(query)
    .build();
};

/** Update a single policy for a customer by its id */
export const actionPutPoliciesByPolicyId = (params: UsePutPoliciesByPolicyIdParams): Action<{
  msg?: string;
} | unknown> => {
    const path = '/api/policies/v1.0/policies/{policyId}'
    .replace('{policyId}', params.policyId.toString());

    const query = {} as Record<string, any>;

    query.dry = params.dry;

    return actionBuilder('PUT', path)
    .queryParams(query)
    .data(params.body)
    .build();
};

export const actionGetStatus = (): Action<unknown> => {
    const path = '/api/policies/v1.0/status';

    const query = {} as Record<string, any>;

    return actionBuilder('GET', path)
    .queryParams(query)
    .build();
};

"
`;
